### PR TITLE
Sum reverse autodiff

### PR DESCRIPTION
**Pull request by gordoncaleb**
This PR:

- Reverse Autodiff with Sum
- Bug fix for elementwise multiply when of is a scalar

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 1]
```diff
@@ -32,7 +38,18 @@ protected DualNumber dualOp(DualNumber dualNumber) {
...
-        throw new UnsupportedOperationException("Reverse mode autodiff not supported for this operation");
+
+        Map<VertexId, DoubleTensor> derivativeOfOutWrtInput = new HashMap<>();
+        for (Map.Entry<VertexId, DoubleTensor> entry : derivativeOfOutputsWithRespectToSelf.asMap().entrySet()) {
+            VertexId of = entry.getKey();
+            DoubleTensor partialOfWrtSelf = entry.getValue();
+
+            int[] outputShape = Arrays.copyOf(partialOfWrtSelf.getShape(), partialOfWrtSelf.getShape().length - getShape().length);
+            DoubleTensor ones = DoubleTensor.ones(TensorShape.concat(outputShape, inputVertex.getShape()));
+            DoubleTensor partial = ones.times(partialOfWrtSelf);
```
Comment (GeorgeNash) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: I've seen this trick of making `ones` of a desired shape to stretch the lower rank tensor. Thoughts on pulling it out into a method? I think it'll make it clearer what's going on. 

```
    @Override
    public DoubleTensor stretch(int[] stretchedShape) {
        //check it's legal stretch shape
        DoubleTensor ones = DoubleTensor.ones(stretchedShape);
        return ones.times(this);
    }
```

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumberTensorTest.java [hunk 2]
```diff
@@ -67,16 +70,17 @@ public void diffWrtScalarOverMultipleMultipliesAndSummation() {
...
-        DoubleVertex prod = A.sum().times(ConstantVertex.of(new double[]{1, 2, 3, 4})).sum();
+        DoubleVertex B = A.sum().times(ConstantVertex.of(new double[]{1, 2, 3, 4})).sum();
 
-        DualNumber dualNumber = prod.getDualNumber();
+        DualNumber dualNumber = B.getDualNumber();
 
         DoubleTensor wrtA = dualNumber.getPartialDerivatives().withRespectTo(A);
 
-        DoubleTensor expectedWrt = DoubleTensor.create(new double[]{4, 8, 12, 16}).reshape(1, 1, 2, 2);
+        //B = 1*(a00 + a01 + a10 + a11) + 2*(a00 + a01 + a10 + a11)+ 3*(a00 + a01 + a10 + a11)+ 4*(a00 + a01 + a10 + a11)
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumberTensorTest.java]: nice!

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java [hunk 3]
```diff
@@ -1,32 +1,125 @@
...
+        DoubleVertex d = a.sum();
+
+        DoubleVertex e = new UniformVertex(new int[]{2, 2}, 0, 10);
+        e.setValue(DoubleTensor.arange(4, 8).reshape(2, 2));
+
+        DoubleVertex f = d.times(e);
+
+        DoubleTensor dfdaForward = f.getDualNumber().getPartialDerivatives().withRespectTo(a);
+
+        PartialDerivatives dfdx = Differentiator.reverseModeAutoDiff(f, a, e);
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java]: nit: can we just have `Differentiator.reverseModeAutoDiff(f, a);` instead - so it's clearer we're just after wrtA

Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java]: Yup. Good point.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 4]
```diff
@@ -32,7 +38,18 @@ protected DualNumber dualOp(DualNumber dualNumber) {
...
-        //TODO
-        throw new UnsupportedOperationException("Reverse mode autodiff not supported for this operation");
+
+        Map<VertexId, DoubleTensor> derivativeOfOutWrtInput = new HashMap<>();
+        for (Map.Entry<VertexId, DoubleTensor> entry : derivativeOfOutputsWithRespectToSelf.asMap().entrySet()) {
+            VertexId of = entry.getKey();
+            DoubleTensor partialOfWrtSelf = entry.getValue();
+
+            int[] outputShape = Arrays.copyOf(partialOfWrtSelf.getShape(), partialOfWrtSelf.getShape().length - getShape().length);
+            DoubleTensor ones = DoubleTensor.ones(TensorShape.concat(outputShape, inputVertex.getShape()));
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: I moved this logic to use the PartialDerivative multiply.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java (hunk 1 above)
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: I don't like giving it a name that's not really well defined in the tensor op world. We could go with a broadcastMultiplyFromLeft and right and that would need to be done with all ops that multiply. That'd be a lot of code and at the moment this only happens in 3 places. I'd say lets revisit it in another PR if it comes up more.

**Pull request merged**